netebpfext/net_ebpf_ext_sock_addr.c

```diff
@@ -784,101 +885,221 @@ _net_ebpf_extension_connection_context_initialize(
+}
+
+static bool
+_net_ebpf_ext_find_and_remove_connection_context(
     uint64_t transport_endpoint_handle, _In_ const bpf_sock_addr_t* sock_addr_ctx)
 {
     KIRQL old_irql;
+    bool entry_found = false;
     net_ebpf_extension_connection_context_t local_connection_context = {0};
-    net_ebpf_extension_connection_context_t* connection_context = NULL;
 
     _net_ebpf_extension_connection_context_initialize(
-        sock_addr_ctx, transport_endpoint_handle, 0, &local_connection_context);
-    old_irql = ExAcquireSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock);
-
-    LIST_ENTRY* list_entry = _net_ebpf_ext_connect_context_list.Flink;
-    while (list_entry != &_net_ebpf_ext_connect_context_list) {
-        net_ebpf_extension_connection_context_t* entry =
-            CONTAINING_RECORD(list_entry, net_ebpf_extension_connection_context_t, list_entry);
-        if (memcmp(
-                &local_connection_context, entry, EBPF_OFFSET_OF(net_ebpf_extension_connection_context_t, timestamp)) ==
-            0) {
-            // Found matching entry. Remove it from the list and return.
-            RemoveEntryList(&entry->list_entry);
-            _net_ebpf_ext_connect_context_count--;
-            connection_context = entry;
-            break;
-        }
-        list_entry = list_entry->Flink;
-    }
+        transport_endpoint_handle, sock_addr_ctx, 0, &local_connection_context);
 
-    ExReleaseSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock, old_irql);
+    old_irql = ExAcquireSpinLockExclusive(&_net_ebpf_ext_sock_addr_blocked_contexts.lock);
+    entry_found = _net_ebpf_ext_find_and_remove_connection_context_locked(&local_connection_context);
+    ExReleaseSpinLockExclusive(&_net_ebpf_ext_sock_addr_blocked_contexts.lock, old_irql);
 
-    NET_EBPF_EXT_RETURN_POINTER(net_ebpf_extension_connection_context_t*, connection_context);
+    return entry_found;
 }
 
-_Requires_exclusive_lock_held_(_net_ebpf_ext_sock_addr_lock) static void _net_ebpf_ext_purge_lru_contexts_under_lock(
-    BOOLEAN delete_all)
+_Requires_exclusive_lock_held_(_net_ebpf_ext_sock_addr_blocked_contexts
+                                   .lock) static void _net_ebpf_ext_purge_blocked_connect_contexts(bool delete_all)
 {
     uint64_t expiry_time = CONVERT_100NS_UNITS_TO_MS(KeQueryInterruptTime()) - EXPIRY_TIME;
 
-    LIST_ENTRY* list_entry = _net_ebpf_ext_connect_context_list.Blink;
-    while (list_entry != &_net_ebpf_ext_connect_context_list) {
+    net_ebpf_extension_connection_context_t* context = NULL;
+    for (context = (net_ebpf_extension_connection_context_t*)RtlEnumerateGenericTableAvl(
+             &_net_ebpf_ext_sock_addr_blocked_contexts.blocked_context_table, TRUE);
+         context != NULL;
+         context = (net_ebpf_extension_connection_context_t*)RtlEnumerateGenericTableAvl(
+             &_net_ebpf_ext_sock_addr_blocked_contexts.blocked_context_table, FALSE)) {
+
+        if (!delete_all && context->timestamp > expiry_time) {
```

Comment:
I think this logic will not work with hash tables and we will need to enumerate all the entries. In case of linked list, the entries were sorted based on the time stamp. Not sure is the same is guaranteed by `RtlEnumerateGenericTableAvl()`

Reply:
Latest commit has the LRU logic added back in. LMK how it looks.